src/Lecture2.hs

```diff
@@ -1,170 +1,277 @@
+-- signatures. If the type signature is not already written, write the
+-- most polymorphic type signature you can.
+--
+-- Unlike exercises to Lecture 1, this module also contains more
+-- challenging exercises. You don't need to solve them to finish the
+-- course but you can if you like challenges :)
 module Lecture2
-    ( -- * Normal
-      lazyProduct
-    , duplicate
-    , removeAt
-    , evenLists
-    , dropSpaces
-
-    , Knight (..)
-    , dragonFight
-
-      -- * Hard
-    , isIncreasing
-    , merge
-    , mergeSort
-
-    , Expr (..)
-    , Variables
-    , EvalError (..)
-    , eval
-    , constantFolding
-    ) where
+  ( -- * Normal
+    lazyProduct,
+    duplicate,
+    removeAt,
+    evenLists,
+    dropSpaces,
+    Knight (..),
+    Dragon (..),
+    DragonType (..),
+    Chest (..),
+    Attack (..),
+    Health (..),
+    Endurance (..),
+    Experience (..),
+    Gold (..),
+    FightResult (..),
+    Treasure (..),
+    Reward (..),
+    dragonFight,
+
+    -- * Hard
+    isIncreasing,
+    merge,
+    mergeSort,
+    Expr (..),
+    Variables,
+    EvalError (..),
+    eval,
+    constantFolding,
+  )
+where
 
 -- VVV If you need to import libraries, do it after this line ... VVV
+import Data.Char (isSpace)
+import Data.Maybe ()
 
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
-
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (x : xs) = x * lazyProduct xs
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate = concatMap (replicate 2)
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt n (x : xs)
+  | n < 0 = (Nothing, x:xs)
+  | n == 0 = (Just x, xs)
+  | otherwise = (ys, x : zs)
+  where
+    (ys, zs) = removeAt (n - 1) xs
```

Comment:
You can move the recursive call into the helper `go` function to avoid comparing with 0 on every single recursive call and do this only once in the beginning 👍🏻